docs/abi.rst

```diff
@@ -0,0 +1,560 @@
+
+.. code-block:: python
+
+    myAddress = abi.make(abi.Address)
+    myBool = abi.make(abi.Bool)
+    myUint64 = abi.make(abi.Uint64)
+    myTuple = abi.make(abi.Tuple3[abi.Address, abi.Bool, abi.Uint64])
+
+    program = Seq(
+        myAddress.set(Txn.sender()),
+        myBool.set(Txn.fee() == Int(0)),
+        myUint64.set(5000),
+        myTuple.set(myAddress, myBool, myUint64)
+    )
+
+Getting Values
+''''''''''''''''''''''
+
+All basic types that represent a single value have a :code:`get()` method, which can be used to extract that value. The supported types and methods are:
+
+* :any:`abi.Uint.get()`, which is used by all :code:`abi.Uint` classes and :code:`abi.Byte`
+* :any:`abi.Bool.get()`
+* :any:`abi.Address.get()`
+* :any:`abi.String.get()`
+
+A brief example is below. Please consult the documentation linked above for each method to learn more about specific usage and behavior.
+
+.. code-block:: python
+
+    @Subroutine(TealType.uint64)
+    def minimum(a: abi.Uint64, b: abi.Uint64) -> Expr:
+        """Return the minimum value of the two arguments."""
+        return (
+            If(a.get() < b.get())
+            .Then(a.get())
+            .Else(b.get())
+        )
+
+Getting Values at Indexes
+''''''''''''''''''''''''''
+
+The types :code:`abi.StaticArray`, :code:`abi.Address`, :code:`abi.DynamicArray`, :code:`abi.String`, and :code:`abi.Tuple` are compound types, meaning they contain other types whose values can be extracted. The :code:`__getitem__` method, accessible by using square brackets to "index into" an object, can be used to extract these values.
+
+The supported methods are:
+
+* :any:`abi.StaticArray.__getitem__`, used for :code:`abi.StaticArray` and :code:`abi.Address`
+* :any:`abi.Array.__getitem__`, used for :code:`abi.DynamicArray` and :code:`abi.String`
+* :any:`abi.Tuple.__getitem__`
+
+Be aware that these methods return a :code:`ComputedValue`, TODO link to Computed Value section
+
+A brief example is below. Please consult the documentation linked above for each method to learn more about specific usage and behavior.
+
+.. code-block:: python
+
+    @Subroutine(TealType.none)
+    def ensure_all_values_greater_than_5(array: abi.StaticArray[abi.Uint64, L[10]]) -> Expr:
+        """This subroutine asserts that every value in the input array is greater than 5."""
+        i = ScratchVar(TealType.uint64)
+        return For(
+            i.store(Int(0)), i.load() < array.length(), i.store(i.load() + Int(1))
+        ).Do(
+            array[i.load()].use(lambda value: Assert(value.get() > Int(5)))
+        )
+
+Limitations
+"""""""""""""""""""""
+
+TODO: explain type size limitations
+
+Reference Types
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some AVM operations require specific values to be placed in the "foreign arrays" of the app call transaction. Reference types allow methods to describe these requirements.
```

Comment:
I suggest giving the reader just a little more here, something like (not in love with how I wrote it, but as an example): 

Many applications require the caller to specify "foreign arrays" when calling the app. These are the blockchain entities (such as accounts, assets, or other applications) that the application will interact with when executing this call. In the ABI, these are called "reference types". Reference types allow methods to describe these requirements.